### PR TITLE
[MI-2505]: Added logic to protect subscriptions notification webhook API and fixed Boards update subscription.

### PR DESCRIPTION
**Pull request by avas27JTG**
Added logic to protect subscriptions notification webhook API and fixed Boards update subscription

Previous PR is https://github.com/Brightscout/mattermost-plugin-azure-devops/pull/18

### CODE REVIEW

**Comment by avas27JTG** [comment]
@ayusht2810 @raghavaggarwal2308 review fixes of 1st pass are done https://github.com/Brightscout/mattermost-plugin-azure-devops/pull/18

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/api_test.go [hunk 1]
```diff
@@ -800,7 +820,7 @@ func TestHandleSubscriptionNotifications(t *testing.T) {
 				return testCase.isValidChannelID
 			})
 
-			req := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/notification?channelID=%s", testCase.channelID), bytes.NewBufferString(testCase.body))
+			req := httptest.NewRequest(http.MethodPost, fmt.Sprintf("/notification?%s=%s&%s=%s", constants.AzureDevopsQueryParamChannelID, testCase.channelID, constants.AzureDevopsQueryParamWebhookSecret, testCase.webhookSecret), bytes.NewBufferString(testCase.body))
```
Comment (ayusht2810) [inline server/plugin/api_test.go:823]: we can use this constant `constants.PathSubscriptionNotifications` here

File: server/plugin/client.go [hunk 2]
```diff
@@ -139,8 +139,14 @@ func (c *client) callFormURLEncoded(url, path, method string, out interface{}, f
 func (c *client) CreateSubscription(body *serializers.CreateSubscriptionRequestPayload, project *serializers.ProjectDetails, channelID, pluginURL, mattermostUserID string) (*serializers.SubscriptionValue, int, error) {
 	subscriptionURL := fmt.Sprintf(constants.CreateSubscription, body.Organization)
 
+	encryptedWebhookSecret, err := c.plugin.Encrypt([]byte(c.plugin.getConfiguration().WebhookSecret), []byte(c.plugin.getConfiguration().EncryptionSecret))
```
Comment (manojmalik20) [inline server/plugin/client.go:142]: What was the need for encrypting the webhook secret?

File: server/plugin/utils.go [hunk 3]
```diff
@@ -343,3 +343,19 @@ func (p *Plugin) GetSubscriptionsForAccessibleChannelsOrProjects(subscriptionLis
 
 	return filteredSubscriptionList, nil
 }
+
+func (p *Plugin) VerifyEncryptedWebhookSecret(received string) (status int, err error) {
+	decodedWebhookSecret, err := p.Decode(received)
+	if err != nil {
+		return http.StatusInternalServerError, errors.New("failed to decode webhook secret")
+	}
```
Comment (manojmalik20) [inline server/plugin/utils.go:351]:
```suggestion
	}
	
```

File: server/plugin/utils.go [hunk 4]
```diff
@@ -343,3 +343,19 @@ func (p *Plugin) GetSubscriptionsForAccessibleChannelsOrProjects(subscriptionLis
...
+	decryptedWebhookSecret, err := p.Decrypt(decodedWebhookSecret, []byte(p.getConfiguration().EncryptionSecret))
+	if err != nil {
+		return http.StatusInternalServerError, errors.New("failed to decrypt webhook secret")
+	}
```
Comment (manojmalik20) [inline server/plugin/utils.go:355]:
```suggestion
	}
	
```

File: server/plugin/utils.go [hunk 5]
```diff
@@ -343,3 +343,19 @@ func (p *Plugin) GetSubscriptionsForAccessibleChannelsOrProjects(subscriptionLis
...
+	if p.getConfiguration().WebhookSecret != string(decryptedWebhookSecret) {
```
Comment (manojmalik20) [inline server/plugin/utils.go:356]: Use something like this - https://github.com/Brightscout/mattermost-plugin-servicenow/blob/953125b99019304f398e30bc2f7e8df3aeba1438/server/plugin/api.go#L149

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -737,13 +738,17 @@ func TestHandleSubscriptionNotifications(t *testing.T) {
 	defer monkey.UnpatchAll()
 	mockAPI := &plugintest.API{}
 	p := setupMockPlugin(mockAPI, nil, nil)
+	p.setConfiguration(&config.Configuration{
```
Comment (manojmalik20) [inline server/plugin/api_test.go:741]: This should be inside the `setupMockPlugin` method.

File: server/plugin/utils.go (hunk 5 above)
Comment (avas27JTG) [reply server/plugin/utils.go:356]: cannot use that directly here as we are receiving as query param and not URL encoded FORM

File: server/plugin/client.go (hunk 2 above)
Comment (avas27JTG) [reply server/plugin/client.go:142]: 1. on Azure we can get this secret only as query-param and when special characters like `+, -, etc` is present then query param breaks so it's better to send and receive an encoded string
1. encryption is done from the security checks perspective if someone steals the webhook token from the Azure portal then that can be misused to post on MM

File: server/plugin/utils.go (hunk 5 above)
Comment (manojmalik20) [reply server/plugin/utils.go:358]: I meant this - https://github.com/Brightscout/mattermost-plugin-servicenow/blob/953125b99019304f398e30bc2f7e8df3aeba1438/server/plugin/api.go#L824

Comment (avas27JTG) [reply server/plugin/utils.go:358]: Tried it, escaped URL query param is not supported by the Azure DevOps, I checked further it gives error as
`failed to create subscription: errorMessage Subscription input 'url' does not match the required data type 'Uri'.`

Comment (avas27JTG) [reply server/plugin/utils.go:358]: As discussed we don't need this, so resolving it

**Comment by avas27JTG** [comment]
@manojmalik20 `strings.EqualFold()` is used for case insensitive comparing of two strings, so I think we should not use this for webhook secret as it could reduce 1 step to strictly check it

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**